File: packages/sdk-py/src/juris_agi_sdk/juris_agi_client.py
"""
JURIS-AGI API Client for Python.
"""

from typing import Any, Callable, Optional

import httpx

from .types import AnalyzeRequest, AnalyzeResponse


class JurisAgiError(Exception):
    """Error from JURIS-AGI API."""

    def __init__(self, message: str, status_code: int):
        super().__init__(message)
        self.status_code = status_code


class JurisAgiClient:
    """
    Client for the JURIS-AGI API.

    Example:
        async with JurisAgiClient(base_url="...") as client:
            result = await client.analyze(deal_id="my-deal")
    """

    def __init__(
        self,
        base_url: str,
        auth_token: Optional[str] = None,
        timeout: float = 60.0,
    ):
        self.base_url = base_url.rstrip("/")
        self.auth_token = auth_token
        self.timeout = timeout
        self._client: Optional[httpx.AsyncClient] = None

    async def __aenter__(self) -> "JurisAgiClient":
        headers = {"Content-Type": "application/json"}
        if self.auth_token:
            headers["Authorization"] = f"Bearer {self.auth_token}"

        self._client = httpx.AsyncClient(
            base_url=self.base_url,
            headers=headers,
            timeout=self.timeout,
        )
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        if self._client:
            await self._client.aclose()
            self._client = None

    @property
    def client(self) -> httpx.AsyncClient:
        if self._client is None:
            raise RuntimeError("Client not initialized. Use 'async with' context.")
        return self._client

    async def _request(self, method: str, path: str, **kwargs) -> dict:
        """Make HTTP request and handle errors."""
        response = await self.client.request(method, path, **kwargs)
        if not response.is_success:
            try:
                error = response.json()
                message = error.get("message", f"HTTP {response.status_code}")
            except Exception:
                message = f"HTTP {response.status_code}"
            raise JurisAgiError(message, response.status_code)
        return response.json()

    async def health(self) -> dict:
        """Health check."""
        return await self._request("GET", "/api/health")

    async def analyze(
        self,
        deal_id: str,
        question: Optional[str] = None,
        claims: Optional[list[dict[str, Any]]] = None,
        **kwargs,
    ) -> AnalyzeResponse:
        """Run JURIS-AGI analysis on a deal."""
        request = AnalyzeRequest(
            deal_id=deal_id,
            question=question,
            claims=claims,
            **kwargs,
        )
        data = await self._request(
            "POST",
            "/api/analyze",
            json=request.model_dump(exclude_none=True),
        )
        return AnalyzeResponse(**data)

    async def analyze_stream(
        self,
        deal_id: str,
        question: Optional[str] = None,
        on_event: Optional[Callable[[str], None]] = None,
        **kwargs,
    ) -> None:
        """Stream analysis results."""
        request = AnalyzeRequest(deal_id=deal_id, question=question, **kwargs)

        async with self.client.stream(
            "POST",
            "/api/analyze/stream",
            json=request.model_dump(exclude_none=True),
        ) as response:
            if not response.is_success:
                raise JurisAgiError(f"HTTP {response.status_code}", response.status_code)

            async for chunk in response.aiter_text():
                if on_event:
                    on_event(chunk)

    async def build_context(
        self,
        deal_id: str,
        question: Optional[str] = None,
        max_claims: int = 100,
        min_confidence: float = 0.5,
    ) -> dict:
        """Build evidence context for analysis."""
        return await self._request(
            "POST",
            "/api/context/build",
            json={
                "deal_id": deal_id,
                "question": question,
                "max_claims": max_claims,
                "min_confidence": min_confidence,
            },
        )

    async def generate_report(
        self,
        analysis_id: str,
        format: str = "markdown",
        include_sections: Optional[list[str]] = None,
    ) -> dict:
        """Generate analysis report."""
        payload: dict[str, Any] = {"analysis_id": analysis_id, "format": format}
        if include_sections:
            payload["include_sections"] = include_sections

        return await self._request("POST", "/api/reports/generate", json=payload)

File: packages/sdk-ts/src/juris-agi-client.ts
/**
 * JURIS-AGI Client
 */

import type {
  AnalyzeRequest,
  AnalyzeResponse,
  BuildContextRequest,
  BuildContextResponse,
  GenerateReportRequest,
  GenerateReportResponse,
} from './types';

export interface JurisAgiConfig {
  baseUrl: string;
  authToken?: string;
  timeout?: number;
}

export class JurisAgiClient {
  private baseUrl: string;
  private authToken?: string;
  private timeout: number;

  constructor(config: JurisAgiConfig) {
    this.baseUrl = config.baseUrl.replace(/\/$/, '');
    this.authToken = config.authToken;
    this.timeout = config.timeout ?? 60000;
  }

  private async fetch<T>(
    path: string,
    options: RequestInit = {}
  ): Promise<T> {
    const url = `${this.baseUrl}${path}`;
    const headers: HeadersInit = {
      'Content-Type': 'application/json',
      ...(this.authToken && { Authorization: `Bearer ${this.authToken}` }),
      ...options.headers,
    };

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);

    try {
      const response = await fetch(url, {
        ...options,
        headers,
        signal: controller.signal,
      });

      if (!response.ok) {
        const error = await response.json().catch(() => ({}));
        throw new JurisAgiError(
          error.message || `HTTP ${response.status}`,
          response.status
        );
      }

      return response.json();
    } finally {
      clearTimeout(timeoutId);
    }
  }

  /**
   * Health check
   */
  async health(): Promise<{ status: string; service: string; version?: string }> {
    return this.fetch('/api/health');
  }

  /**
   * Run analysis on a deal
   */
  async analyze(request: AnalyzeRequest): Promise<AnalyzeResponse> {
    return this.fetch('/api/analyze', {
      method: 'POST',
      body: JSON.stringify(request),
    });
  }

  /**
   * Stream analysis results
   */
  async analyzeStream(
    request: AnalyzeRequest,
    onEvent: (event: string) => void
  ): Promise<void> {
    const url = `${this.baseUrl}/api/analyze/stream`;
    const headers: HeadersInit = {
      'Content-Type': 'application/json',
      ...(this.authToken && { Authorization: `Bearer ${this.authToken}` }),
    };

    const response = await fetch(url, {
      method: 'POST',
      headers,
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      throw new JurisAgiError(`HTTP ${response.status}`, response.status);
    }

    const reader = response.body?.getReader();
    if (!reader) throw new JurisAgiError('No response body', 500);

    const decoder = new TextDecoder();
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      onEvent(decoder.decode(value));
    }
  }

  /**
   * Build evidence context
   */
  async buildContext(request: BuildContextRequest): Promise<BuildContextResponse> {
    return this.fetch('/api/context/build', {
      method: 'POST',
      body: JSON.stringify(request),
    });
  }

  /**
   * Generate report
   */
  async generateReport(request: GenerateReportRequest): Promise<GenerateReportResponse> {
    return this.fetch('/api/reports/generate', {
      method: 'POST',
      body: JSON.stringify(request),
    });
  }
}

export class JurisAgiError extends Error {
  constructor(
    message: string,
    public statusCode: number
  ) {
    super(message);
    this.name = 'JurisAgiError';
  }
}
